For Ember Times No. 119: Writeup on @Pzuraq 's Blog Post on Mut and 2 Way Data Bindings (#312)
ember-learn/ember-blog: source/2019-10-11-the-ember-times-issue-119.md

---

## [A Mutable, Octane Powered Future 🐹🏎](https://www.pzuraq.com/on-mut-and-2-way-binding/)

A handy **helper** has existed in the framework's API since the early days of Ember 1.13. It allows to specify that a certain value - which was passed down by a parent component or controller - **may be mutated** by a child component. This is [the so-called `{{mut}}` helper](https://api.emberjs.com/ember/release/classes/Ember.Templates.helpers/methods/mut?anchor=mut).

Most Ember developers are familiar with the usage of `{{mut}}` from scenarios as the following one:

```handlebars
<!-- app/templates/components/my-component.hbs -->

> <input value={{@passedDownValue}} oninput={{action (mut @passedDownValue) value="target.value"}} />

Reviewer: For small-width resolution, can we break this line into multiple like this?
```suggestion
<input
  value={{@passedDownValue}}
  oninput={{action (mut @passedDownValue) value="target.value"}}
/>
```

```

In combination with the `{{action}}` helper, `{{mut}}` allows developers to create a useful, implicit and template-only **shorthand** for a setter function, that otherwise would require an explicit definition in the component class itself:

```js
// app/components/my-component.js
import Component from '@ember/component';

export default Component.extend({
  actions: {
    updateValue(newValue) {
      this.set('passedDownValue', newValue);
    },
  },
});
```

```handlebars
<!-- app/templates/components/my-component.hbs -->

> <input value={{@passedDownValue}} oninput={{action "updateValue" value="target.value"}} />

Reviewer: If yes to above, then, similarly,
```suggestion
<input
  value={{@passedDownValue}}
  oninput={{action "updateValue" value="target.value"}}
/>
```

```

With [Ember Octane coming soon](https://blog.emberjs.com/2019/08/15/octane-release-plan.html) and the new Glimmer Component API [enforcing unidirectional data flow](https://blog.emberjs.com/2019/03/14/coming-soon-in-ember-octane-part-5.html), many developers might wonder how the `{{mut}}` helper still fits into the overall picture. Are there any issues with using `{{mut}}` in future Octane apps? And are there still plans to [deprecate the helper](https://github.com/emberjs/rfcs/issues/538) or 2-way-binding APIs from the framework in general?

> You will find answers to all of these questions in an [excellent deep-dive into `mut` and 2-way-bindings](https://www.pzuraq.com/on-mut-and-2-way-binding/) by [Chris Garrett (@pzuraq)](https://github.com/pzuraq). This blog post will clarify which data-binding framework APIs will persevere in the foreseeable future (including Ember Octane) and which questions are still open in regards to 2-way binding components commonly used in Ember apps today.

Reviewer: I don't think persevere is the right word here (to me, persevere means endure through difficulties). Maybe remain instead?
```suggestion
You will find answers to all of these questions in an [excellent deep-dive into `mut` and 2-way-bindings](https://www.pzuraq.com/on-mut-and-2-way-binding/) by [Chris Garrett (@pzuraq)](https://github.com/pzuraq). This blog post will clarify which data-binding framework APIs will remain for the foreseeable future (including Ember Octane) and which questions are still open in regards to 2-way binding components commonly used in Ember apps today.
```
